instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py

```diff
@@ -81,15 +81,15 @@ async def hello():
 
 .. code-block:: python
 
-    def server_request_hook(span: Span, scope: dict):
+    def server_request_hook(span: Span, scope: dict[str, Any]):
         if span and span.is_recording():
             span.set_attribute("custom_user_attribute_from_request_hook", "some-value")
 
-    def client_request_hook(span: Span, scope: dict):
+    def client_request_hook(span: Span, scope: dict[str, Any], message: dict[str, Any]):
```

Comment:
> The [call order change here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/eaacb51b22fc88ae426ac5b8c665363e969613de/instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py#L670-L672) basically ensures that the hook could access the ASGI event.

If request hooks were meant to be called right after span creation and before requests are made, should the hook even be allowed to accept the `message` field?

Reply:
I don't think there is even concensus on that, but the existing instrumentations seem to more agree with me: in Django, Requests, `urllib`, `urllib3`, and HTTPX instrumentations, the request hook is able to access the request instance. It's either this or there seems to be no concensus at all. E.g. Flask is a little weird as it receives only the `environ` part of the request instance even though the latter is available.

Also, I could argue that this:

> If request hooks were meant to be called right after span creation and before requests are made

depends heavily on definition of «before request is made». In ASGI instrumentation, we have `server_request_hook` that gets called before any reading and processing happens.

Reply:
Trying to consolidate what was said in [this](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2546#discussion_r1621092953) thread and the this one. I agree we should have a concensus on what these hooks should be doing and what parameters would add the most value/functionality to customers. Currently taken from the docs: 

> This instrumentation supports request and response hooks. These are functions that get called
right after a span is created for a request and right before the span is finished for the response.

This definition doesn't seem too helpful in terms of implementation as well as functionality. If we were to define how hooks would be useful, it probably would be "a place where information pertaining to the current request/response is most present so that I can use that information to run custom logic".

If we agree on this definition, then I see the value of having as much information available for the hooks as possible pertaining to those specific steps. However, I do have one question:

Also, apologies for my inexperience with ASGI, but if we were to make both client* hooks have the `message` and `scope` arguments and we change the ordering of the `client_request_hook` AFTER the actual recieve, what would be a use case in which the `client_request_hook` must be used but the `client_response_hook` can't? What would be the purpose of having both in this case?

Reply:
Great! Thanks for the explanation!